File: z3wire/weave/emit_proto_test.py
"""Tests for the proto emitter."""

import unittest

from z3wire.weave import rdl_pb2
from z3wire.weave.emit_proto import emit_proto
from z3wire.weave.resolver import resolve


class EmitProtoTest(unittest.TestCase):
    def test_simple_struct(self):
        module = rdl_pb2.Module(
            file_prefix="test",
            namespace="test",
            field_pack_order=rdl_pb2.FIELD_PACK_ORDER_LSB_FIRST,
            structs=[
                rdl_pb2.Struct(
                    name="Simple",
                    desc="A simple struct",
                    fields=[
                        rdl_pb2.Field(
                            name="flag",
                            type=rdl_pb2.FieldType(bool=rdl_pb2.BoolType()),
                        ),
                        rdl_pb2.Field(
                            name="value",
                            type=rdl_pb2.FieldType(
                                bitvec=rdl_pb2.BitVecType(width=8)
                            ),
                        ),
                        rdl_pb2.Field(
                            name="pad",
                            type=rdl_pb2.FieldType(
                                bitvec=rdl_pb2.BitVecType(width=7)
                            ),
                            reserved=True,
                        ),
                    ],
                ),
            ],
        )
        resolved = resolve(module)
        output = emit_proto(resolved)

        self.assertIn('syntax = "proto3";', output)
        self.assertIn("package test;", output)
        self.assertIn("message SimpleProto", output)
        self.assertIn("bool flag = 1;", output)
        self.assertIn("uint32 value = 2;", output)
        # Reserved field should be omitted
        self.assertNotIn("pad", output)

    def test_nested_and_signed(self):
        module = rdl_pb2.Module(
            file_prefix="test",
            namespace="example",
            field_pack_order=rdl_pb2.FIELD_PACK_ORDER_LSB_FIRST,
            structs=[
                rdl_pb2.Struct(
                    name="Inner",
                    fields=[
                        rdl_pb2.Field(
                            name="val",
                            type=rdl_pb2.FieldType(
                                bitvec=rdl_pb2.BitVecType(
                                    width=4,
                                    signedness=rdl_pb2.SIGNEDNESS_SIGNED,
                                )
                            ),
                        ),
                    ],
                ),
                rdl_pb2.Struct(
                    name="Outer",
                    fields=[
                        rdl_pb2.Field(
                            name="inner",
                            type=rdl_pb2.FieldType(struct_ref="Inner"),
                        ),
                        rdl_pb2.Field(
                            name="data",
                            type=rdl_pb2.FieldType(
                                bitvec=rdl_pb2.BitVecType(width=4)
                            ),
                            count=3,
                        ),
                    ],
                ),
            ],
        )
        resolved = resolve(module)
        output = emit_proto(resolved)

        self.assertIn("InnerProto", output)
        self.assertIn("sint32 val = 1;", output)
        self.assertIn("InnerProto inner = 1;", output)
        self.assertIn("repeated uint32 data = 2;", output)

    def test_enum_ref_field(self):
        module = rdl_pb2.Module(
            file_prefix="test",
            namespace="test",
            field_pack_order=rdl_pb2.FIELD_PACK_ORDER_LSB_FIRST,
            enums=[
                rdl_pb2.EnumDef(
                    name="Mode",
                    width=2,
                    values=[
                        rdl_pb2.EnumValue(name="kA", value=0),
                        rdl_pb2.EnumValue(name="kB", value=1),
                    ],
                ),
            ],
            structs=[
                rdl_pb2.Struct(
                    name="S",
                    fields=[
                        rdl_pb2.Field(
                            name="mode",
                            type=rdl_pb2.FieldType(enum_ref="Mode"),
                        ),
                    ],
                ),
            ],
        )
        resolved = resolve(module)
        output = emit_proto(resolved)

        self.assertIn("uint32 mode = 1;", output)


if __name__ == "__main__":
    unittest.main()

File: z3wire/weave/emit_proto_test.cc
#include "z3wire/weave/emit_proto.h"

#include <string>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "z3wire/weave/rdl.pb.h"
#include "z3wire/weave/resolver.h"

namespace z3wire::weave {
namespace {

using ::testing::HasSubstr;
using ::testing::Not;

// Helper to create a module with common defaults.
z3wire_rdl::Module MakeModule(const std::string& ns = "test") {
  z3wire_rdl::Module m;
  m.set_file_prefix("test");
  m.set_namespace_(ns);
  m.set_field_pack_order(z3wire_rdl::FIELD_PACK_ORDER_LSB_FIRST);
  return m;
}

TEST(EmitProtoTest, SimpleStruct) {
  auto module = MakeModule();

  auto* s = module.add_structs();
  s->set_name("Simple");
  s->set_desc("A simple struct");

  {
    auto* f = s->add_fields();
    f->set_name("flag");
    f->mutable_type()->mutable_bool_();
  }
  {
    auto* f = s->add_fields();
    f->set_name("value");
    f->mutable_type()->mutable_bitvec()->set_width(8);
  }
  {
    auto* f = s->add_fields();
    f->set_name("pad");
    f->mutable_type()->mutable_bitvec()->set_width(7);
    f->set_reserved(true);
  }

  auto resolved = Resolve(module);
  std::string output = EmitProto(resolved);

  EXPECT_THAT(output, HasSubstr("syntax = \"proto3\";"));
  EXPECT_THAT(output, HasSubstr("package test;"));
  EXPECT_THAT(output, HasSubstr("message SimpleProto"));
  EXPECT_THAT(output, HasSubstr("bool flag = 1;"));
  EXPECT_THAT(output, HasSubstr("uint32 value = 2;"));
  // Reserved field should be omitted.
  EXPECT_THAT(output, Not(HasSubstr("pad")));
}

TEST(EmitProtoTest, NestedAndSigned) {
  auto module = MakeModule("example");

  {
    auto* s = module.add_structs();
    s->set_name("Inner");
    auto* f = s->add_fields();
    f->set_name("val");
    auto* bv = f->mutable_type()->mutable_bitvec();
    bv->set_width(4);
    bv->set_signedness(z3wire_rdl::SIGNEDNESS_SIGNED);
  }
  {
    auto* s = module.add_structs();
    s->set_name("Outer");
    {
      auto* f = s->add_fields();
      f->set_name("inner");
      f->mutable_type()->set_struct_ref("Inner");
    }
    {
      auto* f = s->add_fields();
      f->set_name("data");
      f->mutable_type()->mutable_bitvec()->set_width(4);
      f->set_count(3);
    }
  }

  auto resolved = Resolve(module);
  std::string output = EmitProto(resolved);

  EXPECT_THAT(output, HasSubstr("InnerProto"));
  EXPECT_THAT(output, HasSubstr("sint32 val = 1;"));
  EXPECT_THAT(output, HasSubstr("InnerProto inner = 1;"));
  EXPECT_THAT(output, HasSubstr("repeated uint32 data = 2;"));
}

TEST(EmitProtoTest, EnumRefField) {
  auto module = MakeModule();

  {
    auto* e = module.add_enums();
    e->set_name("Mode");
    e->set_width(2);
    {
      auto* v = e->add_values();
      v->set_name("kA");
      v->set_value(0);
    }
    {
      auto* v = e->add_values();
      v->set_name("kB");
      v->set_value(1);
    }
  }
  {
    auto* s = module.add_structs();
    s->set_name("S");
    auto* f = s->add_fields();
    f->set_name("mode");
    f->mutable_type()->set_enum_ref("Mode");
  }

  auto resolved = Resolve(module);
  std::string output = EmitProto(resolved);

  EXPECT_THAT(output, HasSubstr("uint32 mode = 1;"));
}

}  // namespace
}  // namespace z3wire::weave
